File: iosApp/iosApp/Modules/MnemonicConfirmation ✅/MnemonicConfirmationWireframe.swift
// Created by web3d4v on 12/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultMnemonicConfirmationWireframe {
    private weak var parent: UIViewController?
    private let keyStoreService: KeyStoreService
    private let actionsService: ActionsService
    private let networksService: NetworksService
    
    private weak var vc: UIViewController?

    init(
        _ parent: UIViewController?,
        keyStoreService: KeyStoreService,
        actionsService: ActionsService,
        networksService: NetworksService
    ) {
        self.parent = parent
        self.keyStoreService = keyStoreService
        self.actionsService = actionsService
        self.networksService = networksService
    }
}

extension DefaultMnemonicConfirmationWireframe {

    func present() {
        let vc = wireUp()
        parent?.present(vc, animated: true)
    }
    
    func navigate(with destination: MnemonicConfirmationWireframeDestination) {
        if destination is MnemonicConfirmationWireframeDestination.Dismiss {
            vc?.popOrDismiss()
        }
    }
}

private extension DefaultMnemonicConfirmationWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultMnemonicConfirmationInteractor(
            keyStoreService: keyStoreService,
            actionsService: actionsService,
            networksService: networksService
        )
        let vc: MnemonicConfirmationViewController = UIStoryboard(.mnemonicConfirmation).instantiate()
        let presenter = DefaultMnemonicConfirmationPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor
        )
        vc.presenter = presenter
        let nc = NavigationController(rootViewController: vc)
        self.vc = nc
        return nc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/mnemonicconfirmation/MnemonicConfirmationWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicconfirmation

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3walletcore.app.App
import com.sonsofcrypto.web3walletcore.modules.mnemonicConfirmation.DefaultMnemonicConfirmationInteractor
import com.sonsofcrypto.web3walletcore.modules.mnemonicConfirmation.DefaultMnemonicConfirmationPresenter
import com.sonsofcrypto.web3walletcore.modules.mnemonicConfirmation.MnemonicConfirmationWireframe
import com.sonsofcrypto.web3walletcore.modules.mnemonicConfirmation.MnemonicConfirmationWireframeDestination
import com.sonsofcrypto.web3walletcore.modules.mnemonicImport.MnemonicImportWireframeDestination
import com.sonsofcrypto.web3walletcore.services.actions.ActionsService
import com.sonsofcrypto.web3walletcore.services.mnemonic.MnemonicService

class DefaultMnemonicConfirmationWireframe(
    private val parent: WeakRef<Fragment>?,
    private val keyStoreService: KeyStoreService,
    private val actionsService: ActionsService,
    private val mnemonicService: MnemonicService,
): MnemonicConfirmationWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.present(fragment, animated = true)
    }

    override fun navigate(destination: MnemonicConfirmationWireframeDestination) {
        when (destination) {
            is MnemonicConfirmationWireframeDestination.Dismiss -> {
                parent?.navigationFragment?.popOrDismiss()
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = MnemonicConfirmationFragment()
        val interactor = DefaultMnemonicConfirmationInteractor(
            keyStoreService,
            actionsService,
            mnemonicService,
        )
        val presenter = DefaultMnemonicConfirmationPresenter(
            WeakRef(view),
            this,
            interactor,
        )
        view.presenter = presenter
        return view
    }
}